### PR TITLE
CI: add mypy to strict type check

### PR DESCRIPTION
**Pull request by beauxq**
## What is this fixing or adding?

This is an addition to https://github.com/ArchipelagoMW/Archipelago/pull/3121 with some previous discussion here https://github.com/ArchipelagoMW/Archipelago/pull/3107

There are some errors that mypy catches and pyright doesn't, and some errors that pyright catches and mypy doesn't.
So we can catch more problems if we check with both.

The one Python script `.github/type_check.py` will show output from both mypy and pyright.

One of the biggest issues we've found with mypy in AP is this bug: https://github.com/python/mypy/issues/10506
After lots of battling with it in https://github.com/ArchipelagoMW/Archipelago/pull/2173 and https://github.com/ArchipelagoMW/Archipelago/pull/2899, I think we worked around it so it only needs a `# type: ignore` for `default = "random"`
I think that will only happen with a few options in worlds, and this check is opt-in from world maintainers.

## How was this tested?

submitting this PR, because thar's the only way to test github actions

### CODE REVIEW

**Comment by beauxq** [comment]
A minor issue I found is this:
```python
def foo(option_class: Union[Type[NumericOption], Type[FreeText]]) -> None:
    print(option_class.name_lookup.values())
```
reports: "error: Access to generic class variables is ambiguous  [misc]"
This should be pretty rare, because it only happens with the union of multiple options, and not with 1 base type `Type[Option[int]]`

File: .github/type_check.py [hunk 1]
```diff
@@ -1,15 +1,60 @@
...
+        print(f"{e} - Is mypy installed?")
+        exit(1)
+    return mypy_result.returncode
+
+
+if __name__ == "__main__":
+    # mypy is first because of its --install-types feature
+    mypy_ret = mypy()
+    pyright_ret = pyright()
+    exit(mypy_ret and pyright_ret)
```
Comment (silasary) [inline .github/type_check.py:60]: Considering that exit codes are "0 = good", this should probably be an or.

**Review by silasary — CHANGES_REQUESTED** [review]

**Review by silasary — APPROVED** [review]

File: .github/type_check.py [hunk 2]
```diff
@@ -1,15 +1,60 @@
...
+    except FileNotFoundError as e:
+        print(f"{e} - Is mypy installed?")
+        exit(1)
+    return mypy_result.returncode
+
+
+if __name__ == "__main__":
+    # mypy is first because of its --install-types feature
+    mypy_ret = mypy()
+    pyright_ret = pyright()
```
Comment (black-sliver) [inline .github/type_check.py:59]: I suggest renaming those so it's easier to understand when only looking at a part of the code.
```suggestion
    mypy_ret = run_mypy()
    pyright_ret = run_pyright()
```

File: .github/type_check.py [hunk 3]
```diff
@@ -1,15 +1,60 @@
...
-command = ("pyright", "-p", str(config))
-print(" ".join(command))
 
-try:
-    result = subprocess.run(command)
-except FileNotFoundError as e:
-    print(f"{e} - Is pyright installed?")
-    exit(1)
+def pyright() -> int:
+    """ returns process return code """
```
Comment (black-sliver) [inline .github/type_check.py:11]: Afaik the generic name is exit code or exist status, since it's the integer you pass to exit.
```suggestion
def run_pyright() -> int:
    """ returns process exit code """
```

File: .github/type_check.py [hunk 4]
```diff
@@ -1,15 +1,60 @@
...
+    try:
+        pyright_result = subprocess.run(command)
+    except FileNotFoundError as e:
+        print(f"{e} - Is pyright installed?")
+        exit(1)
+    return pyright_result.returncode
+
+
+def mypy() -> int:
+    """ returns process return code """
```
Comment (black-sliver) [inline .github/type_check.py:24]: As above
```suggestion
def run_mypy() -> int:
    """ returns process exit code """
```

File: .github/type_check.py [hunk 5]
```diff
@@ -1,15 +1,60 @@
...
+    with open(config) as config_file:
+        config_data: Union[Dict[str, Any], Any] = json.load(config_file)
```
Comment (black-sliver) [inline .github/type_check.py:26]: Shouldn't mypy use a separate list of files, so as the maintainer of a code snippet can decide which one they think should be applied for their code (assuming they are not 1:1 compatible)?

Comment (beauxq) [reply .github/type_check.py:26]: We can't really isolate the checks to a specific list of files.
The list of files we give (to either mypy or pyright) is not a list of files to check, It's a list of files to report errors in. Both type checkers follow imports to get type information from other files. So the type information in any file is going to affect both checks.
If someone thinks they're only using mypy, because the file they're changing is only on the mypy list, that change will still affect the pyright check, even though the file is not on the pyright list.

The main focus of this is core, and worlds are opt-in.
If this PR is accepted (I don't consider it decided yet whether we want this), If a world maintainer wants to only use one, then they shouldn't opt in to this system, and should just check with their own workflow.

Comment (beauxq) [reply .github/type_check.py:26]: You shouldn't think of this as a linter, checking stylistic things.
You should think of it as a static type checker, just like in a C++ or Rust compiler.

I think it wouldn't make sense to say:
"For this C++ file `x.cpp` I only care whether clang's type checker passes, but for `y.cpp` I only care whether gcc's type checker passes."
We would care about everything passing, in one compiler, or the other compiler, or we want to support both compilers.

I think it would be good to support multiple static type checkers, just like a C++ project that wants to support multiple compilers.

Comment (black-sliver) [reply .github/type_check.py:26]: That argument makes sense, yeah.

That being said, I honestly don't know how we (or who) decide(s) if we want to merge this or not.

**Comment by black-sliver** [comment]
I guess I would like for more people to chime in and say if they think this is good or bad. It's hard for me to decide if this is an upgrade or not.

In my mind, having both allows more proper placement of typing or `# type: ignore` to ensure compatibility with both mypy and pyright, however pinning both in CI will also give more opportunities for it to be out of line with what people use locally.

In other words,
if the CI succeeds we can be fairly certain that running any type checker locally will work, but
if we get a fix in either one, we still can not "use" it until it's fixed in both and the pins got updated.

**Comment by black-sliver** [comment]
@NewSoupVi do you have an opinion on this?

**Comment by beauxq** [comment]
The more I think about this, the more I question whether it's a good idea.

In addition to all the mypy bugs, mypy is too slow to use for in-editor warnings which makes it more difficult to work with.

going to draft, might go to closed

**Comment by beauxq** [comment]
> One of the biggest issues we've found with mypy in AP is this bug: https://github.com/python/mypy/issues/10506
After lots of battling with it in https://github.com/ArchipelagoMW/Archipelago/pull/2173 and https://github.com/ArchipelagoMW/Archipelago/pull/2899, I think we worked around it so it only needs a # type: ignore for default = "random"

And it looks like mypy has decided this bug is "as designed"...
https://github.com/python/mypy/issues/10506#issuecomment-2644236553
